components/asana/asana.app.js

```diff
@@ -0,0 +1,251 @@
+          return {
+            label: project.name,
+            value: project.gid,
+          };
+        });
+      },
+    },
+    taskIds: {
+      type: "string[]",
+      label: "Tasks",
+      async options(opts) {
+        const tasks = await this.getTasks(opts.projectId);
+        return tasks.map((task) => {
+          return { label: task.name, value: task.gid };
+        });
+      },
+    },
+    organizationId: {
+      type: "string",
+      label: "Organization",
+      optional: false,
+      async options(opts) {
+        const organizations = await this.getOrganizations();
+        return organizations.map((organization) => {
+          return {
+            label: organization.name,
+            value: organization.gid,
+          };
+        });
+      },
+    },
+  },
+  methods: {
+    async _getHeaders() {
+      return {
+        Accept: "application/json",
+        Authorization: `Bearer ${this.$auth.oauth_access_token}`,
+      };
+    },
+    async _getAuthorizationHeader({ data, method, url, headers }) {
+      const token = {
+        key: this.$auth.oauth_access_token,
+        secret: this.$auth.oauth_refresh_token,
+      };
+      return await axios({
+        method: "POST",
+        url: `https://app.asana.com/api/1.0/webhooks`,
+        data: data.body,
+        headers,
+      });
+    },
+    async _makeWebhookRequest(config) {
+      if (!config.headers) config.headers = {};
+      const authorization = await this._getAuthorizationHeader(config);
+      config.headers.authorization = authorization;
+      try {
+        await axios(config);
+      } catch (err) {
+        console.log(err);
+      }
+      return authorization.data;
+    },
+    async createHook(body) {
+      const resp = await this._makeWebhookRequest({
+        method: "post",
+        url: `https://app.asana.com/api/1.0/webhooks`,
+        headers: {
+          "Content-Type": "applicaton/json",
+          Accept: "application/json",
+          Authorization: `Bearer ${this.$auth.oauth_access_token}`,
+        },
+        data: {
+          body,
+        },
+      });
+      return resp;
+    },
+    async deleteHook(hookId) {
+      const config = {
+        method: "delete",
+        url: `https://app.asana.com/api/1.0/webhooks/${hookId}`,
+        headers: await this._getHeaders(),
+      };
+      try {
+        await axios(config);
+      } catch (err) {
+        console.log(err);
+      }
+    },
+    async verifyAsanaWebhookRequest(request) {
+      let secret = this.$auth.oauth_refresh_token;
+      var base64Digest = function (s) {
+        return crypto.createHmac("sha1", secret).update(s).digest("base64");
+      };
+      var content = JSON.stringify(request.body);
+      var doubleHash = base64Digest(content);
+      var headerHash = request.headers["x-hook-secret"];
+      return doubleHash == headerHash;
+    },
+    async getWorkspace(workspaceId) {
+      const workspace = await axios.get(
```

Comment:
Similar to Github and other apps, I think you could abstract some of the HTTP request logic into a `_makeRequest` method so that you don't have to pass the full URL or headers in every case. Instead, you can include those attributes of the config in that single method and then just pass the path to the `_makeRequest` method. You're already close to this with the `_makeWebhookRequest` method, so I think there's the opportunity to consolidate that further.